ddl/index.go

```diff
@@ -324,171 +326,261 @@ func (d *ddl) onDropIndex(t *meta.Meta, job *model.Job) error {
+	if ret.count > 0 {
+		ret.doneHandle = idxRecords[ret.count-1].handle
+	}
+	// Be sure to do this operation only once.
+	if !handleInfo.isRetry {
+		// Notice to start the next batch operation.
+		batchOpInfo.nextCh <- ret.doneHandle
+		// Record the last handle.
+		// Ensure that the handle scope of the batch doesn't change,
+		// even if the transaction retries it can't effect the other batches.
+		handleInfo.endHandle = ret.doneHandle
+		handleInfo.isRetry = true
+	}
+	if ret.count == 0 {
+		return nil, ret
+	}
+
+	cols := t.Cols()
+	idxInfo := batchOpInfo.tblIndex.Meta()
+	for i, idxRecord := range idxRecords {
+		rowMap, err := tablecodec.DecodeRow(rawRecords[i], batchOpInfo.colMap)
+		if err != nil {
+			ret.err = errors.Trace(err)
+			return nil, ret
+		}
+		idxVal := make([]types.Datum, 0, len(idxInfo.Columns))
+		for _, v := range idxInfo.Columns {
+			col := cols[v.Offset]
+			idxVal = append(idxVal, rowMap[col.ID])
+		}
+		idxRecord.vals = idxVal
+	}
+	return idxRecords, ret
 }
 
-const defaultBatchCnt = 1024
-const defaultSmallBatchCnt = 128
+const (
+	defaultBatchCnt      = 1024
+	defaultSmallBatchCnt = 128
+	defaultSmallBatches  = 16
+)
+
+// batchResult is the result of the batch.
+type batchResult struct {
+	count      int   // The number of records that has been proceed in the batch.
+	doneHandle int64 // This is the last reorg handle that has been processed.
+	err        error
+}
+
+type batchRetSlice []*batchResult
+
+func (b batchRetSlice) Len() int           { return len(b) }
+func (b batchRetSlice) Less(i, j int) bool { return b[i].doneHandle < b[j].doneHandle }
+func (b batchRetSlice) Swap(i, j int)      { b[i], b[j] = b[j], b[i] }
+
+// indexRecord is the record information of an index.
+type indexRecord struct {
+	handle int64
+	key    []byte        // It's used to lock a record. Record it to reduce the encoding time.
+	vals   []types.Datum // It's the index values.
+}
+
+// indexBatchOpInfo records the information that is needed in the batch process.
+type indexBatchOpInfo struct {
+	tblIndex   table.Index
+	colMap     map[int64]*types.FieldType // It's the index columns map.
+	batchRetCh chan *batchResult          // Get the results of all batches.
+	nextCh     chan int64                 // It notifies to start the next batch.
+}
 
 // How to add index in reorganization state?
-//  1. Generate a snapshot with special version.
-//  2. Traverse the snapshot, get every row in the table.
-//  3. For one row, if the row has been already deleted, skip to next row.
-//  4. If not deleted, check whether index has existed, if existed, skip to next row.
-//  5. If index doesn't exist, create the index and then continue to handle next row.
+// Concurrently process defaultSmallBatches tasks. Each task deals with a handle interval of the index record.
+// The handle interval is defaultSmallBatchCnt.
+// Although the length of each handle interval is controllable, but the range of the handle value can't be expected,
```

Comment:
Because each handle range depends on the previous one, it's necessary to obtain the handle range sequentially.